### ISSUE TITLE
Engine url_helpers went missing from 4.1 -> 4.2

### ISSUE THREAD

**Issue by dancinglightning** [rails/rails#22662]
I have written an engine https://github.com/rubyclerks/office_clerk
that i have been using in an app  https://github.com/rubyclerks/sales_clerk for 2 years already.

The engine was not namespaced but everything worked in 4.1. By working i mean url_helpers were included and so i could use an engine path like clerks_path in the templates.

I have now upgraded and had to change _all paths_ to reference the engines name, so the example becomes office.clerks_path

This brings issues with it: form_helpers don't work the same anymore, will_paginate doesn't find paths, best_in_place doesn't find paths. I managed to hack all those and everything works again, but it is very unsatisfactory.

So i went back to investigate. The OfficeClerk::Engine.routes.url_helper is working after the routes have been defined. But it is _not working_ in the engine initializer, even with an after block. And they are not working in the OfficeController when they are included.
And even including the helpers into the ApplicationController at the end of the routes, does not help.
btw: rake routes does give the clerks_path without the scope, even that does not work.

I have tried to dig deep, reading rails source for a day now, but i can't say i understand it fully. One of the last initialisers re-inits the routes, i sort of suspect that. Something with the load order having changed slightly and since the mechanism is based on ruby include, it does not get refreshed after the routes refresh. Something like that ?

I even tried with 2.2.4 and the current 4-2-stable branch incase #22322 had any effect, but no luck.

Since the engine worked in 4.1 i feel this is a bug. I'd be happy to help track it down, but without help i am at the end of my tether.

PS: I am seeing almost double time spent in views after upgrade, maybe related?

**Comment by dancinglightning** [comment]
So, more digging, and i managed to fix it in the simplest of ways
   
     class OfficeController < ApplicationController
       helper  OfficeClerk::Engine.routes.url_helpers

now that seems obvious off course. I thought in all that testing i had checked a newly generated isolated and not isolated engine, and they didn't include such lines. I though. And the way i understood things such explicit things should not be needed, but maybe wrong.

Do close if i'm wrong. Or investigate if not. Or not :-)

**Comment by eileencodes** [comment]
Can you bisect 4.1 to 4.2 and find what commit removed the `url_helpers`? Also if you could provide a basic engine that demonstrates the problem, that would help us narrow down the issue. Thanks!

**Comment by dancinglightning** [comment]
Hi,
nice tool, didn't know that bisect. Gave it a spin and it came out with what  knew anyway.
This commit https://github.com/rubyclerks/office_clerk/commit/ffa74966bb016b65f75a8da0db2b96f5d66eb9a3

update my gemfile from rails 4.1.14 to 4.2.5. There's a little fluff in there because i tried to generate a new  rails app and overlaid that on the existing code to see if some config / engine changes where made.
I just commited that and then start on fixing it with a bind to branch from before if needed later.

So basically it is the rails update to 4.2, not any of my code, as far as i see.

The engine is there for download/cloning/ investigating. It is a basic engine, and up there is the commit that broke it. I don't know what else i can say/do. 
I'm sorry i don't have the time to do the debugging, if that's what you're asking. I have spent a day on this and i luckily managed to fix it. I am a one man shop and have to move on.
But i'll answer questions if you want to investigate.

**Comment by eileencodes** [comment]
Hey @dancinglightning I meant to use git bisect on Rails to find what commit in Rails broken your app :smile: Let me know if that's not possible.

**Comment by dancinglightning** [comment]
Hi @eileencodes 

I don't think it's quite as straightforward. Neither do i know if the code before that commit worked with 4.2 nor if it would still work with 4.1

When doing the bisect on rails and my constant engine, the tests will require the bundle to be bundled and that will nail the rails version. So i really don't know.

So now i neither have the time, nor know how to do that :-)

**Comment by eileencodes** [comment]
Can you make a small app the reproduces the problem? Then I can bisect Rails.

**Comment by dancinglightning** [comment]
Ok, i thought this would be quick so i did it
https://github.com/dancinglightning/blorgh

but it raises other issues, possibly of documentation and you will really have to take it from here.
The thing is that i'm sure if you use --mountable all works. But i didn't, neither now nor then and things seem to go awry.

There's only three commit in there and if you go to test/dummy and do rails s
and go to /blorgh , you'll see some of the weirdness.

it says No route matches [GET] "/blorgh"
but the routes include:         blorgh_engine_path		/blorgh	Blorgh::Engine

also the articles from the scaffold are not mounted?

As i said more issues. Possibly lack of testing for --full. 
Sorry i have no more time.
